File: test/scripts/context-switching.js
#!/usr/bin/env node
/**
 * ANTI-PATTERN: Excessive context switching (Worker threads)
 * Resume: "Reduce Context Switching" - More worker threads than CPU cores
 * causes constant rotation and scheduler overhead.
 *
 * Run: node context-switching.js
 * Measure: vmstat 1, pidstat -w
 */

const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');
const os = require('os');

const NUM_WORKERS = 64;  // Way more than cores
const ITERATIONS = 10_000_000;

if (isMainThread) {
  const numCores = os.cpus().length;
  console.log('CPU cores:', numCores);
  console.log(`Creating ${NUM_WORKERS} workers (>> cores) -> context switch storm`);
  console.log('Run: vmstat 1  or  pidstat -w -p <pid> 1');

  const start = Date.now();
  const workers = [];

  for (let i = 0; i < NUM_WORKERS; i++) {
    workers.push(new Promise((resolve) => {
      const w = new Worker(__filename, { workerData: { id: i } });
      w.on('message', () => {});
      w.on('exit', resolve);
    }));
  }

  Promise.all(workers).then(() => {
    const elapsed = (Date.now() - start) / 1000;
    console.log(`Done in ${elapsed.toFixed(2)}s (with ${NUM_WORKERS} workers)`);
    console.log('Compare with NUM_WORKERS=numCores for better performance');
  });
} else {
  let total = 0;
  for (let i = 0; i < ITERATIONS; i++) {
    total += i * (i + 1);
  }
  parentPort.postMessage(total);
}

File: test/scripts/context-switching.py
#!/usr/bin/env python3
"""
ANTI-PATTERN: Excessive context switching
Resume: "Reduce Context Switching" - More threads than CPU cores causes
constant rotation, cache eviction, and scheduler overhead.

Run: python context-switching.py
Measure: vmstat 1, pidstat -w
"""

import threading
import multiprocessing
import time

NUM_THREADS = 500  # Way more than typical CPU cores (8-16)
ITERATIONS = 100_000


def cpu_busy_work():
    """Small CPU-bound task that gets interrupted often."""
    total = 0
    for i in range(ITERATIONS):
        total += i * (i + 1)  # Pure CPU work
    return total


def main():
    num_cores = multiprocessing.cpu_count()
    print(f"CPU cores: {num_cores}")
    print(f"Creating {NUM_THREADS} threads (>> cores) -> context switch storm")
    print("Run: vmstat 1  or  pidstat -w -p $(pgrep -f context-switching) 1")

    start = time.perf_counter()
    threads = []
    for _ in range(NUM_THREADS):
        t = threading.Thread(target=cpu_busy_work)
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    elapsed = time.perf_counter() - start
    print(f"Done in {elapsed:.2f}s (with {NUM_THREADS} threads)")
    print("Compare with: NUM_THREADS=8 for ~num_cores -> much faster")


if __name__ == "__main__":
    main()
